src/contracts/fungible/cache/cache.rs

```diff
@@ -27,6 +29,18 @@ pub trait Cache {
     fn has_asset(&self, id: ContractId) -> Result<bool, Self::Error>;
     fn add_asset(&mut self, asset: Asset) -> Result<bool, Self::Error>;
     fn remove_asset(&mut self, id: ContractId) -> Result<bool, Self::Error>;
+
+    /// Return the map of Utxo-Allocation_amount for a given asset
+    fn utxo_allocation_map(
+        &self,
+        contract_id: ContractId,
+    ) -> Result<BTreeMap<bitcoin::OutPoint, AtomicValue>, Self::Error>;
+
+    /// Returns the map of Asset-Allocation_amount for a given Outpoint
+    fn asset_allocation_map(
+        &self,
+        utxo: &bitcoin::OutPoint,
+    ) -> Result<BTreeMap<String, u32>, CacheError>;
```

Comment:
Same here as above, we need `HashSet` and `Vec<u32>` as values.

Regarding the keys, we need to index against ContractId, not String.

Also, why don't we use `AtomicAmount` here and use `u32`?